aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -217,12 +238,20 @@ void TensorIterator::compute_types() {
           op.dtype = op.tensor.scalar_type();
         } else {
           op.device = common_device;
-          op.dtype = common_dtype;
+          if (compute_common_dtype_only_for_inputs && op.is_output) {
+            op.dtype = op.tensor.scalar_type();
+          } else {
+            op.dtype = common_dtype;
+          }
         }
       }
 
-      validate_dtype(op, common_dtype, ninputs());
-      maybe_promote_common_dtype(op, common_dtype);
+      if (compute_common_dtype_strategy_ == CommonDTypeStrategy::COMPUTE_ALL) {
+        validate_dtype(op, common_dtype, ninputs());
+      }
+      if ((compute_common_dtype_strategy_ == CommonDTypeStrategy::COMPUTE_ALL) || !op.is_output) {
+        maybe_promote_common_dtype(op, common_dtype);
```

Comment:
Should we do this even if `!op.is_output && compute_common_dtype_strategy == CommonDTypeStrategy::None`?

Reply:
I changed this to: !compute_common_dtype_only_for_inputs || !op.is_output

Reply:
My comment still holds: if the operand is an input and `compute_common_dtype_strategy == CommonDTypeStrategy::None`, this code still attempts to promote to the common dtype. Is this the expected behavior?

Reply:
> Should we do this even if `!op.is_output && compute_common_dtype_strategy == CommonDTypeStrategy::None`?

I changed https://github.com/pytorch/pytorch/pull/25593#discussion_r322966199 to:
```
if (!compute_common_dtype_only_for_inputs || !op.is_output) {
        maybe_promote_common_dtype(op, common_dtype);
```
Because we already inside "if" statement with (missing_dtypes || compute_common_dtype), so we always want to promote common dtype, 
except when compute_common_dtype_only_for_inputs == CommonDTypeStrategy::COMPUTE_INPUTS and op is output, 

```
!(compute_common_dtype_only_for_inputs && op.is_output) = !compute_common_dtype_only_for_inputs || !op.is_output
```

Reply:
Oh I see, that is my bad for not noticing. The branching levels of this function go pretty deep, we should really clean it up sometime.